crates/wash-cli/src/dev.rs

```diff
@@ -85,33 +105,117 @@ impl Drop for HostSubprocess {
-    NatsMessagingProvider,
-    /// wasmCloud blobstore (normally corresponds to `wasi:blobstore`)
-    BlobstoreFsProvider,
-    /// Custom provider of some interface (whether a provider or component),
-    /// including the necessary information to identify it as a dependency.
+/// Keys that index the list of dependencies in a [`ProjectDeps`]
+///
+/// # Examples
+///
+/// ```
+/// let project_key = ProjectDependencyKey::Project {
+///   name: "http-hello-world".into(),
+///   imports: vec![ ("wasi".into(), "http".into(), "incoming-handler".into(), None) ],
+///   exports: vec![ ("wasi".into(), "http".into(), "incoming-handler".into(), None) ],
+///   in_workspace: None,
+/// };
+///
+/// let workspace_key = ProjectDependencyKey::Workspace; // alternatively ProjectDependencyKey::default()
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+enum ProjectDependencyKey {
     #[allow(unused)]
-    Custom {
-        import_interfaces: Vec<String>,
-        export_interfaces: Vec<String>,
+    RootWorkspace { name: String, path: PathBuf },
+    /// Identifies a nested workspace inside the root workspace
+    ///
+    /// Workspaces are hierarchical: they may contain one or more projects *or* other workspaces,
+    /// with the path structure of workspaces being the arbiter of which workspaces are above others.
+    ///
+    /// Only one workspace can be the *top-most* workspace, in that it contains
+    /// all other workspaces and projects.
+    #[allow(unused)]
+    Workspace {
         name: String,
-        image_ref: String,
+        path: PathBuf,
+        root: bool,
     },
+    /// Identifies a project inside the root workspace
+    Project { name: String, path: PathBuf },
+}
+
+impl ProjectDependencyKey {
+    /// Create a [`ProjectDependencyKey`] from the name of a project
+    ///
+    /// The supplied `project_dir` must be a folder containing a `wasmcloud.toml`
+    fn from_project(name: &str, project_dir: impl AsRef<Path>) -> Result<Self> {
+        Ok(Self::Project {
+            name: name.into(),
+            path: project_dir.as_ref().into(),
+        })
+    }
+}
+
+/// Specification for a single dependency in a given project
+///
+/// [`DependencySpec`]s are normally gleaned from some source of project metadata, for example:
+///
+/// - dependency overrides in a project-level `wasmcloud.toml`
+/// - dependency overrides in a workspace-level `wasmcloud.toml`
+/// - WIT interface of a project
+///
+/// A `DependencySpec` represents a single dependency in the project, categorized into what part it is expected
+/// to play in in fulfilling WIT interfaces.
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum DependencySpec {
+    /// A dependency that receives invocations (ex. `keyvalue-nats` receiving a `wasi:keyvalue/get`)
+    Receives(DependencySpecInner),
+    /// A dependency that performs invocations (ex. `http-server` invoking a component's `wasi:http/incoming-handler` export)
+    Invokes(DependencySpecInner),
+}
```

Comment:
I don't dislike this terminology, but it is yet-another set of names for what we call export/import and target/source. What do you think about sticking with the WIT terminology?

```suggestion
enum DependencySpec {
    /// A dependency that receives invocations (ex. `keyvalue-nats` receiving a `wasi:keyvalue/get`)
    Exports(DependencySpecInner),
    /// A dependency that performs invocations (ex. `http-server` invoking a component's `wasi:http/incoming-handler` export)
    Imports(DependencySpecInner),
}
```

Reply:
I don't think we've really settled on a perfect thing to call these directions, so I figured sticking with what's already required may be the best move. Open to conversation here